File: mobile/lib/services/projection_snapshot_service.dart
import 'package:flutter/foundation.dart' show kIsWeb;
import 'package:hive_flutter/hive_flutter.dart';
import '../models/contact.dart';
import '../models/transaction.dart';
import '../models/event.dart';
import 'state_builder.dart';
import 'event_store_service.dart';

/// Projection snapshot model stored in Hive
class ProjectionSnapshot {
  final AppState state;
  final DateTime snapshotTimestamp;
  final String lastEventId;
  final int eventCount;
  final int snapshotIndex;

  ProjectionSnapshot({
    required this.state,
    required this.snapshotTimestamp,
    required this.lastEventId,
    required this.eventCount,
    required this.snapshotIndex,
  });

  Map<String, dynamic> toJson() => {
        'contacts': state.contacts.map((c) => c.toJson()).toList(),
        'transactions': state.transactions.map((t) => t.toJson()).toList(),
        'snapshotTimestamp': snapshotTimestamp.toIso8601String(),
        'lastEventId': lastEventId,
        'eventCount': eventCount,
        'snapshotIndex': snapshotIndex,
      };

  factory ProjectionSnapshot.fromJson(Map<String, dynamic> json) {
    final contacts = (json['contacts'] as List)
        .map((c) => Contact.fromJson(c as Map<String, dynamic>))
        .toList();
    final transactions = (json['transactions'] as List)
        .map((t) => Transaction.fromJson(t as Map<String, dynamic>))
        .toList();

    return ProjectionSnapshot(
      state: AppState(
        contacts: contacts,
        transactions: transactions,
        lastBuiltAt: DateTime.parse(json['snapshotTimestamp'] as String),
      ),
      snapshotTimestamp: DateTime.parse(json['snapshotTimestamp'] as String),
      lastEventId: json['lastEventId'] as String,
      eventCount: json['eventCount'] as int,
      snapshotIndex: json['snapshotIndex'] as int,
    );
  }
}

/// Projection Snapshot Service
/// Manages projection snapshots for efficient state rebuilding
class ProjectionSnapshotService {
  static const String boxName = 'projection_snapshots';
  static const int maxSnapshots = 5; // Keep last 5 snapshots
  static const int snapshotInterval = 10; // Create snapshot every 10 events

  /// Save a snapshot
  static Future<void> saveSnapshot(
    AppState state,
    String lastEventId,
    int eventCount,
  ) async {
    if (kIsWeb) return;

    try {
      final box = await Hive.openBox<Map>(boxName);

      // Get current snapshot index
      int nextIndex = 0;
      if (box.isNotEmpty) {
        final lastSnapshot = box.values.last;
        if (lastSnapshot['snapshotIndex'] != null) {
          nextIndex = (lastSnapshot['snapshotIndex'] as int) + 1;
        }
      }

      final snapshot = ProjectionSnapshot(
        state: state,
        snapshotTimestamp: DateTime.now(),
        lastEventId: lastEventId,
        eventCount: eventCount,
        snapshotIndex: nextIndex,
      );

      // Save snapshot with index as key
      await box.put(nextIndex, snapshot.toJson());

      // Cleanup old snapshots
      await cleanupOldSnapshots();

      print('✅ Saved projection snapshot #$nextIndex (event count: $eventCount)');
    } catch (e) {
      print('Error saving projection snapshot: $e');
    }
  }

  /// Get snapshot before a specific event
  /// Returns the most recent snapshot where the snapshot's last event timestamp is before the target event timestamp
  static Future<ProjectionSnapshot?> getSnapshotBeforeEvent(String eventId) async {
    if (kIsWeb) return null;

    try {
      final box = await Hive.openBox<Map>(boxName);
      if (box.isEmpty) return null;

      // Get the target event to find its timestamp
      final targetEvent = await EventStoreService.getEvent(eventId);
      if (targetEvent == null) return null;

      // Get all snapshots sorted by index (descending - most recent first)
      final snapshots = box.values.toList()
        ..sort((a, b) => (b['snapshotIndex'] as int).compareTo(a['snapshotIndex'] as int));

      // Find the most recent snapshot where the snapshot's last event timestamp is before target event timestamp
      for (final snapshotData in snapshots) {
        final snapshot = ProjectionSnapshot.fromJson(Map<String, dynamic>.from(snapshotData));
        
        // Get the snapshot's last event to compare timestamps
        final snapshotLastEvent = await EventStoreService.getEvent(snapshot.lastEventId);
        if (snapshotLastEvent != null) {
          // Return snapshot if its last event is before the target event
          if (snapshotLastEvent.timestamp.isBefore(targetEvent.timestamp)) {
            return snapshot;
          }
        }
      }

      return null;
    } catch (e) {
      print('⚠️ Could not find snapshot before event (this is normal if no snapshots exist): $e');
      return null;
    }
  }

  /// Get the latest snapshot
  static Future<ProjectionSnapshot?> getLatestSnapshot() async {
    if (kIsWeb) return null;

    try {
      final box = await Hive.openBox<Map>(boxName);
      if (box.isEmpty) return null;

      // Get snapshot with highest index
      final snapshots = box.values.toList()
        ..sort((a, b) => (b['snapshotIndex'] as int).compareTo(a['snapshotIndex'] as int));

      if (snapshots.isEmpty) return null;

      return ProjectionSnapshot.fromJson(Map<String, dynamic>.from(snapshots.first));
    } catch (e) {
      print('Error getting latest snapshot: $e');
      return null;
    }
  }

  /// Cleanup old snapshots, keeping only the last maxSnapshots
  static Future<void> cleanupOldSnapshots() async {
    if (kIsWeb) return;

    try {
      final box = await Hive.openBox<Map>(boxName);
      if (box.length <= maxSnapshots) return;

      // Get all snapshots sorted by index (ascending - oldest first)
      final snapshots = box.values.toList()
        ..sort((a, b) => (a['snapshotIndex'] as int).compareTo(b['snapshotIndex'] as int));

      // Delete oldest snapshots, keeping only the last maxSnapshots
      final toDelete = snapshots.length - maxSnapshots;
      for (int i = 0; i < toDelete; i++) {
        final index = snapshots[i]['snapshotIndex'] as int;
        await box.delete(index);
      }

      print('🧹 Cleaned up ${toDelete} old snapshots, kept $maxSnapshots');
    } catch (e) {
      print('Error cleaning up snapshots: $e');
    }
  }

  /// Check if we should create a snapshot based on event count
  static bool shouldCreateSnapshot(int eventCount) {
    return eventCount % snapshotInterval == 0;
  }

  /// Build state from snapshot and events after snapshot
  static Future<AppState?> buildStateFromSnapshot(
    ProjectionSnapshot? snapshot,
    List<Event> eventsAfterSnapshot,
  ) async {
    if (snapshot == null) return null;

    // Filter out undone events and UNDO events
    final undoneEventIds = <String>{};
    for (final event in eventsAfterSnapshot) {
      if (event.eventType == 'UNDO') {
        final undoneEventId = event.eventData['undone_event_id'] as String?;
        if (undoneEventId != null) {
          undoneEventIds.add(undoneEventId);
        }
      }
    }

    final filteredEvents = eventsAfterSnapshot.where((event) {
      if (event.eventType == 'UNDO') return false;
      if (undoneEventIds.contains(event.id)) return false;
      return true;
    }).toList();

    // Apply filtered events to snapshot state
    return StateBuilder.applyEvents(snapshot.state, filteredEvents);
  }
}

File: backend/rust-api/src/services/projection_snapshot_service.rs
use sqlx::{PgPool, Row, postgres::PgRow, FromRow};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ProjectionSnapshot {
    pub id: i64,
    pub snapshot_index: i64,
    pub last_event_id: i64,
    pub event_count: i64,
    pub contacts_snapshot: serde_json::Value,
    pub transactions_snapshot: serde_json::Value,
    pub created_at: chrono::NaiveDateTime,
}

impl<'r> FromRow<'r, PgRow> for ProjectionSnapshot {
    fn from_row(row: &'r PgRow) -> Result<Self, sqlx::Error> {
        Ok(ProjectionSnapshot {
            id: row.try_get("id")?,
            snapshot_index: row.try_get("snapshot_index")?,
            last_event_id: row.try_get("last_event_id")?,
            event_count: row.try_get("event_count")?,
            contacts_snapshot: row.try_get("contacts_snapshot")?,
            transactions_snapshot: row.try_get("transactions_snapshot")?,
            created_at: row.try_get("created_at")?,
        })
    }
}

const MAX_SNAPSHOTS: i64 = 5;
const SNAPSHOT_INTERVAL: i64 = 10;

/// Save a projection snapshot
pub async fn save_snapshot(
    pool: &PgPool,
    last_event_id: i64,
    event_count: i64,
    contacts_snapshot: serde_json::Value,
    transactions_snapshot: serde_json::Value,
    wallet_id: uuid::Uuid,
) -> Result<(), sqlx::Error> {
    // Get next snapshot index for this wallet
    let next_index = sqlx::query_scalar::<_, Option<i64>>(
        "SELECT COALESCE(MAX(snapshot_index), -1) + 1 FROM projection_snapshots WHERE wallet_id = $1"
    )
    .bind(wallet_id)
    .fetch_one(pool)
    .await?;

    let next_index = next_index.unwrap_or(0);

    // Insert snapshot
    sqlx::query(
        r#"
        INSERT INTO projection_snapshots 
        (snapshot_index, last_event_id, event_count, contacts_snapshot, transactions_snapshot, wallet_id)
        VALUES ($1, $2, $3, $4, $5, $6)
        "#
    )
    .bind(next_index)
    .bind(last_event_id)
    .bind(event_count)
    .bind(contacts_snapshot)
    .bind(transactions_snapshot)
    .bind(wallet_id)
    .execute(pool)
    .await?;

    tracing::info!("Saved projection snapshot #{} (event count: {})", next_index, event_count);

    // Cleanup old snapshots for this wallet
    cleanup_old_snapshots(pool, wallet_id).await?;

    Ok(())
}

/// Get snapshot before a specific event ID
/// Returns the most recent snapshot where last_event_id < event_id
pub async fn get_snapshot_before_event(
    pool: &PgPool,
    event_id: i64,
    wallet_id: uuid::Uuid,
) -> Result<Option<ProjectionSnapshot>, sqlx::Error> {
    let snapshot = sqlx::query_as::<_, ProjectionSnapshot>(
        r#"
        SELECT id, snapshot_index, last_event_id, event_count, 
               contacts_snapshot, transactions_snapshot, created_at
        FROM projection_snapshots
        WHERE last_event_id < $1 AND wallet_id = $2
        ORDER BY snapshot_index DESC
        LIMIT 1
        "#
    )
    .bind(event_id)
    .bind(wallet_id)
    .fetch_optional(pool)
    .await?;

    Ok(snapshot)
}

/// Get the latest snapshot for a wallet
#[allow(dead_code)] // Reserved for future snapshot functionality
pub async fn get_latest_snapshot(
    pool: &PgPool,
    wallet_id: uuid::Uuid,
) -> Result<Option<ProjectionSnapshot>, sqlx::Error> {
    let snapshot = sqlx::query_as::<_, ProjectionSnapshot>(
        r#"
        SELECT id, snapshot_index, last_event_id, event_count,
               contacts_snapshot, transactions_snapshot, created_at
        FROM projection_snapshots
        WHERE wallet_id = $1
        ORDER BY snapshot_index DESC
        LIMIT 1
        "#
    )
    .bind(wallet_id)
    .fetch_optional(pool)
    .await?;

    Ok(snapshot)
}

/// Cleanup old snapshots, keeping only the last MAX_SNAPSHOTS for a wallet
pub async fn cleanup_old_snapshots(pool: &PgPool, wallet_id: uuid::Uuid) -> Result<(), sqlx::Error> {
    // Get count of snapshots for this wallet
    let count: i64 = sqlx::query_scalar("SELECT COUNT(*) FROM projection_snapshots WHERE wallet_id = $1")
        .bind(wallet_id)
        .fetch_one(pool)
        .await?;

    if count <= MAX_SNAPSHOTS {
        return Ok(());
    }

    // Delete oldest snapshots, keeping only the last MAX_SNAPSHOTS for this wallet
    let to_delete = count - MAX_SNAPSHOTS;
    sqlx::query(
        r#"
        DELETE FROM projection_snapshots
        WHERE wallet_id = $1 AND snapshot_index IN (
            SELECT snapshot_index
            FROM projection_snapshots
            WHERE wallet_id = $1
            ORDER BY snapshot_index ASC
            LIMIT $2
        )
        "#
    )
    .bind(wallet_id)
    .bind(to_delete)
    .execute(pool)
    .await?;

    tracing::info!("Cleaned up {} old snapshots, kept {}", to_delete, MAX_SNAPSHOTS);

    Ok(())
}

/// Check if we should create a snapshot based on event count
pub fn should_create_snapshot(event_count: i64) -> bool {
    event_count % SNAPSHOT_INTERVAL == 0
}

/// Get event ID from events table by event_id UUID
#[allow(dead_code)] // Reserved for future event lookup functionality
pub async fn get_event_db_id(
    pool: &PgPool,
    event_id: Uuid,
) -> Result<Option<i64>, sqlx::Error> {
    let id = sqlx::query_scalar::<_, Option<i64>>(
        "SELECT id FROM events WHERE event_id = $1"
    )
    .bind(event_id)
    .fetch_optional(pool)
    .await?;

    Ok(id.flatten())
}

/// Get snapshot with event_count less than target_count
/// Returns the most recent snapshot where event_count < target_count
pub async fn get_snapshot_before_event_count(
    pool: &PgPool,
    target_count: i64,
    wallet_id: uuid::Uuid,
) -> Result<Option<ProjectionSnapshot>, sqlx::Error> {
    let snapshot = sqlx::query_as::<_, ProjectionSnapshot>(
        r#"
        SELECT id, snapshot_index, last_event_id, event_count, 
               contacts_snapshot, transactions_snapshot, created_at
        FROM projection_snapshots
        WHERE event_count < $1 AND wallet_id = $2
        ORDER BY snapshot_index DESC
        LIMIT 1
        "#
    )
    .bind(target_count)
    .bind(wallet_id)
    .fetch_optional(pool)
    .await?;

    Ok(snapshot)
}
